lib.php

```diff
@@ -0,0 +1,76 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Library functions for custom content type module
+ *
+ * @package     mod_cms
+ * @author      Marcus Boon<marcus@catalyst-au.net>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+use core_course\local\entity\content_item;
+use mod_cms\local\lib;
+
+/**
+ * Returns whether a feature is supported or not.
+ *
+ * @uses FEATURE_IDNUMBER
+ * @uses FEATURE_GROUPS
+ * @uses FEATURE_GROUPINGS
+ * @uses FEATURE_MOD_INTRO
+ * @uses FEATURE_COMPLETION_TRACKS_VIEWS
+ * @uses FEATURE_GRADE_HAS_GRADE
+ * @uses FEATURE_GRADE_OUTCOMES
+ * @param string $feature FEATURE_xx constant for requested feature
+ *
+ * @return bool|null
+ */
+function cms_supports($feature) {
+    switch($feature) {
+        case FEATURE_IDNUMBER:
+        case FEATURE_MOD_INTRO:
+        case FEATURE_BACKUP_MOODLE2:
+        case FEATURE_NO_VIEW_LINK:
+            return true;
+        case FEATURE_GROUPS:
+        case FEATURE_GROUPINGS:
+        case FEATURE_COMPLETION_TRACKS_VIEWS:
+        case FEATURE_GRADE_HAS_GRADE:
+        case FEATURE_GRADE_OUTCOMES:
+            return false;
+        case FEATURE_MOD_ARCHETYPE:
+            return MOD_ARCHETYPE_RESOURCE;
+        case FEATURE_MOD_PURPOSE:
+            return MOD_PURPOSE_CONTENT;
+        default:
+            return null;
+    }
+}
+
+/**
+ * Obtains a list of defined content types to be included in the activity chooser panel.
+ *
+ * @param content_item $defaultmodulecontentitem
+ * @param stdClass $user
+ * @param stdClass $course
+ * @return array
+ */
+function cms_get_course_content_items(content_item $defaultmodulecontentitem, \stdClass $user,
+    \stdClass $course) {
+    return lib::get_course_content_items($defaultmodulecontentitem, $user, $course);
```

Comment:
Belay that, for some reason when I pulled your version, it didn't have the `use mod_cms\local\lib;` I added that manually and it looks like it's working now
![image](https://github.com/catalyst/moodle-mod_cms/assets/3231180/193c07e2-4a24-4e79-bd7f-51c7482625c1)

However, now we need to define a form so that we can add an instance of the custom content activity as per the second AC. Right now it's exploding with:
```
No formslib form description file found for this activity.

[More information about this error](https://docs.moodle.org/401/en/error/moodle/noformdesc)
error
Error
Debug info:
Error code: noformdesc
error
Error
Stack trace:

    line 148 of /course/modedit.php: moodle_exception thrown

error
Error
Output buffer: <div class="notifytiny debuggingmessage" data-rel="debugging">The module cms does not define the standard capability mod/cms:addinstance<ul style="text-align: left" data-rel="backtrace"><li>line 1953 of /course/lib.php: call to debugging()</li><li>line 501 of /course/modlib.php: call to course_allowed_module()</li><li>line 843 of /course/modlib.php: call to can_add_moduleinfo()</li><li>line 72 of /course/modedit.php: call to prepare_new_moduleinfo_data()</li></ul></div>
```

Reply:
done

Reply:
When you edit the instance that was added it's showing an error now
```
Invalid get_string() identifier: 'pluginadministration' or component 'cms'. Perhaps you are missing $string['pluginadministration'] = ''; in mod/cms/lang/en/cms.php?

    line 356 of /lib/classes/string_manager_standard.php: call to debugging()
    line 7409 of /lib/moodlelib.php: call to core_string_manager_standard->get_string()
    line 4814 of /lib/navigationlib.php: call to get_string()
    line 4342 of /lib/navigationlib.php: call to settings_navigation->load_module_settings()
    line 833 of /lib/pagelib.php: call to settings_navigation->initialise()
    line 962 of /lib/pagelib.php: call to moodle_page->magic_get_settingsnav()
    line 225 of /lib/classes/navigation/views/secondary.php: call to moodle_page->__get()
    line 890 of /lib/pagelib.php: call to core\navigation\views\secondary->initialise()
    line 962 of /lib/pagelib.php: call to moodle_page->magic_get_secondarynav()
    line 72 of /theme/monash/layout/drawers.php: call to moodle_page->__get()
    line 1477 of /lib/outputrenderers.php: call to include()
    line 1403 of /lib/outputrenderers.php: call to core_renderer->render_page_layout()
    line 207 of /course/modedit.php: call to core_renderer->header()
```

Reply:
And then when you try and save after edit it explodes
```
Exception - Call to undefined function cms_update_instance()

[More information about this error](https://docs.moodle.org/401/en/error/moodle/generalexceptionmessage)
error
Error
Debug info:
Error code: generalexceptionmessage
error
Error
Stack trace:

    line 647 of /course/modlib.php: Error thrown
    line 168 of /course/modedit.php: call to update_moduleinfo()

error
Error
Output buffer: <br /> <font size='1'><table class='xdebug-error xe-notice' dir='ltr' border='1' cellspacing='0' cellpadding='1'> <tr><th align='left' bgcolor='#f57900' colspan="5"><span style='background-color: #cc0000; color: #fce94f; font-size: x-large;'>( ! )</span> Notice: Undefined property: stdClass::$introeditor in /var/www/site/course/modlib.php on line <i>631</i></th></tr> <tr><th align='left' bgcolor='#e9b96e' colspan='5'>Call Stack</th></tr> <tr><th align='center' bgcolor='#eeeeec'>#</th><th align='left' bgcolor='#eeeeec'>Time</th><th align='left' bgcolor='#eeeeec'>Memory</th><th align='left' bgcolor='#eeeeec'>Function</th><th align='left' bgcolor='#eeeeec'>Location</th></tr> <tr><td bgcolor='#eeeeec' align='center'>1</td><td bgcolor='#eeeeec' align='center'>0.0003</td><td bgcolor='#eeeeec' align='right'>373992</td><td bgcolor='#eeeeec'>{main}( )</td><td title='/var/www/site/course/modedit.php' bgcolor='#eeeeec'>.../modedit.php<b>:</b>0</td></tr> <tr><td bgcolor='#eeeeec' align='center'>2</td><td bgcolor='#eeeeec' align='center'>0.1560</td><td bgcolor='#eeeeec' align='right'>7672456</td><td bgcolor='#eeeeec'>update_moduleinfo( )</td><td title='/var/www/site/course/modedit.php' bgcolor='#eeeeec'>.../modedit.php<b>:</b>168</td></tr> </table></font> <br /> <font size='1'><table class='xdebug-error xe-notice' dir='ltr' border='1' cellspacing='0' cellpadding='1'> <tr><th align='left' bgcolor='#f57900' colspan="5"><span style='background-color: #cc0000; color: #fce94f; font-size: x-large;'>( ! )</span> Notice: Trying to access array offset on value of type null in /var/www/site/course/modlib.php on line <i>631</i></th></tr> <tr><th align='left' bgcolor='#e9b96e' colspan='5'>Call Stack</th></tr> <tr><th align='center' bgcolor='#eeeeec'>#</th><th align='left' bgcolor='#eeeeec'>Time</th><th align='left' bgcolor='#eeeeec'>Memory</th><th align='left' bgcolor='#eeeeec'>Function</th><th align='left' bgcolor='#eeeeec'>Location</th></tr> <tr><td bgcolor='#eeeeec' align='center'>1</td><td bgcolor='#eeeeec' align='center'>0.0003</td><td bgcolor='#eeeeec' align='right'>373992</td><td bgcolor='#eeeeec'>{main}( )</td><td title='/var/www/site/course/modedit.php' bgcolor='#eeeeec'>.../modedit.php<b>:</b>0</td></tr> <tr><td bgcolor='#eeeeec' align='center'>2</td><td bgcolor='#eeeeec' align='center'>0.1560</td><td bgcolor='#eeeeec' align='right'>7672456</td><td bgcolor='#eeeeec'>update_moduleinfo( )</td><td title='/var/www/site/course/modedit.php' bgcolor='#eeeeec'>.../modedit.php<b>:</b>168</td></tr> </table></font>
```